packages/ckeditor5-code-block/src/converters.js

```diff
@@ -250,3 +250,59 @@ export function dataViewToModelTextNewlinesInsertion() {
 		data.modelCursor = position;
 	};
 }
+
+/**
+ * A view-to-model converter that handles orphan text nodes (white spaces, new lines, etc.)
+ * that surround `<code>` inside `<pre>`.
+ *
+ * Sample input:
+ *
+ *		// White spaces
+ *		<pre> <code>foo()</code> </pre>
+ *
+ *		// White spaces
+ *		<pre>      <code>foo()</code>      </pre>
+ *
+ *		// White spaces
+ *		<pre>			<code>foo()</code>			</pre>
+ *
+ *		// New lines
+ *		<pre>
+ *			<code>foo()</code>
+ *		</pre>
+ *
+ *		// Redundant text
+ *		<pre>ABC<code>foo()</code>DEF</pre>
+ *
+ *		// Redundant elements
+ *		<pre><span>ABC</span><code>foo()</code><b>DEF</b></pre>
+ *
+ * Unified output for each case:
+ *
+ *		<codeBlock language="plaintext">foo()</codeBlock>
+ *
+ * @returns {Function} Returns a conversion callback.
+ */
+export function dataViewToModelOrphanNodeConsumer() {
+	return ( evt, data, { consumable } ) => {
+		// Do nothing, when not inside `<pre>`.
+		if ( !data.viewItem.parent.is( 'element', 'pre' ) ) {
+			return;
+		}
+
+		// Do nothing to `<code>` inside `<pre>`. There's a dedicated `dataViewToModelCodeBlockInsertion()`
+		// converter just for that.
+		if ( data.viewItem.is( 'element', 'code' ) ) {
+			return;
+		}
+
+		const consumables = data.viewItem.is( 'element' ) ? { name: true } : null;
+
+		// Do nothing, if the node has already been converted.
+		if ( !consumable.test( data.viewItem, consumables ) ) {
+			return;
+		}
+
+		consumable.consume( data.viewItem, consumables );
```

Comment:
I don't like that this converter could potentially consume children of a `<pre>` that was not meant to become a code block (e.g. because it has no `<code>`, or just whatever). There could be other features upcasting `<pre>...</pre>` working next to `CodeBlock`.

At this moment, I don't see how could I check whether this pre will become a code block. The default converter for code blocks listens to `element:code`. I thought I could at least look for siblings of consumed orphans here and locate `<code>` and check if it was consumed (better than nothing) but for orphans preceding `<code>`, `consumable.test( siblingCodeElement )` will say it has not been consumed yet because this is too early.

Reply:
We decided to go with checking if the given pre will end up as a code block before trying to strip text nodes. This is the easiest way and matches the list feature (at least partly).